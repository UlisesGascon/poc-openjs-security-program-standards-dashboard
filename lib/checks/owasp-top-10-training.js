import { getConfig } from '../../config'
import { isDateWithinPolicy } from '../utils'

const { generalExpirationPolicy } = getConfig()

export function checkOwaspTop10Training (projects, checks) {
  checks.owaspTop10Training = {
    name: 'OWASP Top 10 Training',
    description: 'At least One Primary Maintainer has taken TBD Training on OWASP Top 10 or Equivalent',
    level: {
      incubating: 'expected',
      graduated: 'expected',
      retiring: 'expected'
    },
    url: 'https://openjs-security-program-standards.netlify.app/details/item-1',
    projects: {}
  }

  for (const project of projects) {
    checks.owaspTop10Training.projects[project.name] = {}
    const trainingDetails = project.additionalInformation?.owaspTop10Training || {}

    const criteria = trainingDetails.globalAchievement &&
            isDateWithinPolicy(trainingDetails.confirmationDate, generalExpirationPolicy)

    if (criteria) {
      checks.owaspTop10Training.projects[project.name].globalAchievement = true
      checks.owaspTop10Training.projects[project.name].rationale = trainingDetails.rationale
    } else {
      checks.owaspTop10Training.projects[project.name].globalAchievement = false
      checks.owaspTop10Training.projects[project.name].rationale = 'The team has not completed the required training or requires an update.'
    }
  }
}
