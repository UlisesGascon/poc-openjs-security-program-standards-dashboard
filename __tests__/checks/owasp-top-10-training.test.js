import { checkOwaspTop10Training } from '../../lib/checks/owasp-top-10-training.js'
import { isDateWithinPolicy } from '../../lib/utils/index.js'

jest.mock('../../lib/utils/index.js')

describe('checkOwaspTop10Training', () => {
  let projects
  let checks
  let tasks
  let alerts

  beforeEach(() => {
    projects = [
      {
        name: 'Project1',
        additionalInformation: {
          owaspTop10Training: {
            globalAchievement: true,
            rationale: 'Training completed',
            confirmationDate: new Date().toISOString()
          }
        }
      },
      {
        name: 'Project2',
        additionalInformation: {
          owaspTop10Training: {
            globalAchievement: false,
            rationale: 'Training not completed',
            confirmationDate: new Date().toISOString()
          }
        }
      }
    ]
    checks = {}
    tasks = {}
    alerts = {}
    isDateWithinPolicy.mockReturnValue(true)
  })

  it('should update checks with project compliance data', () => {
    checkOwaspTop10Training({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with no additional information', () => {
    projects = [
      {
        name: 'Project3'
      }
    ]
    checkOwaspTop10Training({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })

  it('should handle projects with expired training', () => {
    projects = [
      {
        name: 'Project4',
        additionalInformation: {
          owaspTop10Training: {
            globalAchievement: true,
            rationale: 'Training completed',
            confirmationDate: new Date().toISOString()
          }
        }
      }
    ]
    isDateWithinPolicy.mockReturnValue(false)
    checkOwaspTop10Training({ projects, checks, tasks, alerts })
    expect(checks).toMatchSnapshot()
    expect(tasks).toMatchSnapshot()
    expect(alerts).toMatchSnapshot()
  })
})
