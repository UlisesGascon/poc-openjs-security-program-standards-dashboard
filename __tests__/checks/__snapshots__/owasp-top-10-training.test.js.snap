// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkOwaspTop10Training should handle projects with expired training 1`] = `
{
  "owaspTop10Training": {
    "description": "At least One Primary Maintainer has taken TBD Training on OWASP Top 10 or Equivalent",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "OWASP Top 10 Training",
    "projects": {
      "Project4": {
        "globalAchievement": false,
        "rationale": "The team has not completed the required training or requires an update.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-1",
  },
}
`;

exports[`checkOwaspTop10Training should handle projects with no additional information 1`] = `
{
  "owaspTop10Training": {
    "description": "At least One Primary Maintainer has taken TBD Training on OWASP Top 10 or Equivalent",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "OWASP Top 10 Training",
    "projects": {
      "Project3": {
        "globalAchievement": false,
        "rationale": "The team has not completed the required training or requires an update.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-1",
  },
}
`;

exports[`checkOwaspTop10Training should update checks with project compliance data 1`] = `
{
  "owaspTop10Training": {
    "description": "At least One Primary Maintainer has taken TBD Training on OWASP Top 10 or Equivalent",
    "level": {
      "graduated": "expected",
      "incubating": "expected",
      "retiring": "expected",
    },
    "name": "OWASP Top 10 Training",
    "projects": {
      "Project1": {
        "globalAchievement": true,
        "rationale": "Training completed",
      },
      "Project2": {
        "globalAchievement": false,
        "rationale": "The team has not completed the required training or requires an update.",
      },
    },
    "url": "https://openjs-security-program-standards.netlify.app/details/item-1",
  },
}
`;
